File: cracking-the-coding-interview/ctci-find-the-running-median.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
  public static float getMedian(PriorityQueue<Integer> lowers, PriorityQueue<Integer> highers) {
    if (lowers.size() == highers.size()) {
      return ((float) (lowers.peek() + highers.peek()) ) / 2.0f;
    } else if (lowers.size() > highers.size()) {
      return lowers.peek();
    } else {
      return highers.peek();
    }
  }

  public static void rebalance(PriorityQueue<Integer> lowers, PriorityQueue<Integer> highers) {
    if (lowers.size() != highers.size()) {
      PriorityQueue biggerHeap = lowers.size() > highers.size() ? lowers : highers;
      PriorityQueue smallerHeap = lowers.size() < highers.size() ? lowers : highers;

      if ((biggerHeap.size() - smallerHeap.size()) >= 2) {
        smallerHeap.add(biggerHeap.poll());
      }
    }
  }

  public static void addElement(int data, PriorityQueue<Integer> lowers, PriorityQueue<Integer> highers) {
    if (lowers.size() == 0 || data < lowers.peek()) {
      lowers.add(data);
    } else {
      highers.add(data);
    }
  }

  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();
    PriorityQueue lowers = new PriorityQueue<Integer>((a, b)->b-a);
    PriorityQueue highers = new PriorityQueue<Integer>();
    for(int a_i=0; a_i < n; a_i++){
      int data = in.nextInt();
      addElement(data, lowers, highers);
      rebalance(lowers, highers);
      System.out.println(getMedian(lowers, highers));
    }
  }
}

File: cracking-the-coding-interview/ctci-find-the-running-median.py
#!/bin/python3

import sys
import heapq

def addElement(data, lowers, highers):
  if (len(lowers) == 0 or data < abs(lowers[0])):
    # Push -data to make it a max heap
    heapq.heappush(lowers, -data);
  else:
    # Min heap
    heapq.heappush(highers, data);

def rebalance(lowers, highers):
  biggerHeap = lowers if len(lowers) > len(highers) else highers
  smallerHeap = lowers if len(lowers) < len(highers) else highers

  if len(lowers) != len(highers) and (len(biggerHeap) - len(smallerHeap)) >= 2:
    data = heapq.heappop(biggerHeap)
    heapq.heappush(smallerHeap, -data)

def findMedian(lowers, highers):
  if len(lowers) == len(highers):
    return float(abs(lowers[0]) + highers[0]) / 2.0
  elif len(lowers) > len(highers):
    return float(abs(lowers[0]))
  else:
    return float(highers[0])

n = int(input().strip())
a_i = 0
medians = []
lowers = []
highers = []
for a_i in range(n):
  a_t = int(input().strip())
  addElement(a_t, lowers, highers);
  rebalance(lowers, highers)
  medians.append(findMedian(lowers, highers))

for median in medians:
  print(median)
